File: tools/project/templates/cpp/app_core_compute.cpp
// Tellusim Compute Application Template

#include <TellusimApp.h>
#include <core/TellusimCore.h>
#include <math/TellusimMath.h>
#include <platform/TellusimPlatforms.h>

/*
 */
using namespace Tellusim;

/*
 */
class Application {
		
	public:
		
		// constructor initializing the application with command-line arguments
		Application(int32_t argc, char **argv) : app(argc, argv) {
			
		}
		
		// destructor cleans up resources
		virtual ~Application() {
			
			// finish context
			if(window) window.finish();
		}
		
		// create function responsible for initialization
		virtual bool create() {
			
			// create app
			if(!app.create()) {
				TS_LOG(Error, "Application::create(): can't create App\n");
				return false;
			}
			
			// create window
			if(!create_window()) return false;
			
			// create resources
			if(!create_resources()) return false;
			
			// additional initialization code can go here
			
			return true;
		}
		
		// update function for logic updates
		virtual bool update() {
			
			// your update logic here
			
			return true;
		}
		
		// dispatch function to handle the compute commands
		virtual bool dispatch(Compute &compute) {
			
			// resize surface texture
			uint32_t width = window.getWidth();
			uint32_t height = window.getHeight();
			if(!surface || surface.getWidth() != width || surface.getHeight() != height) {
				device.releaseTexture(surface);
				surface = device.createTexture2D(FormatRGBAu8n, width, height, Texture::FlagSurface);
				if(!surface) {
					TS_LOG(Error, "Application::dispatch(): can't create Texture\n");
					return false;
				}
			}
			
			// compute uniform parameters
			struct ComputeParameters {
				float32_t time;
			};
			
			ComputeParameters parameters;
			parameters.time = (float32_t)Time::seconds();
			
			// dispatch kernel
			compute.setKernel(kernel);
			compute.setUniform(0, parameters);
			compute.setSurfaceTexture(0, surface);
			compute.dispatch(surface);
			compute.barrier(surface);
			
			// your dispatch code here
			
			return true;
		}
		
		// render function to handle the drawing commands
		virtual bool render(Command &command) {
			
			// draw fullscreen triangle
			command.setPipeline(pipeline);
			command.setTexture(0, surface);
			command.drawArrays(3);
			
			// your rendering code here
			
			return true;
		}
		
		// main run loop
		virtual bool run() {
			
			// run application
			if(window) return window.run([this]() {
				return render_window();
			});
			
			return true;
		}
		
	private:
		
		// helper function to create and initialize the Window
		virtual bool create_window() {
			
			// create Window
			window = Window(app.getPlatform(), app.getDevice());
			if(!window || !window.setSize(app.getWidth(), app.getHeight()) ||
				!window.create(String(window.getPlatformName()) + " @NAME@") || !window.setHidden(false)) {
				TS_LOG(Error, "Application::create_window(): can't create Window\n");
				return false;
			}
			
			// set Window callbacks
			window.setKeyboardPressedCallback([this](uint32_t key, uint32_t code) {
				if(key == Window::KeyF4 && window.getKeyboardKey(Window::KeyAlt)) window.stop();
				if(key == Window::KeyEsc) window.stop();
			});
			window.setCloseClickedCallback([&]() { window.stop(); });
			
			// create Device
			device = Device(window);
			if(!device) {
				TS_LOG(Error, "Application::create_window(): can't create Device\n");
				return false;
			}
			
			// create Target
			target = device.createTarget(window);
			if(!target) {
				TS_LOG(Error, "Application::create_window(): can't create Target\n");
				return false;
			}
			
			return true;
		}
		
		// helper function to create and load resources
		virtual bool create_resources() {
			
			// create kernel
			kernel = device.createKernel().setSurfaces(1).setUniforms(1);
			if(!kernel.loadShaderGLSL("main.shader", "COMPUTE_SHADER=1")) {
				TS_LOG(Error, "Application::create_resources(): can't load Compute shader\n");
				return false;
			}
			if(!kernel.create()) {
				TS_LOG(Error, "Application::create_resources(): can't create Kernel\n");
				return false;
			}
			
			// create pipeline
			pipeline = device.createPipeline();
			pipeline.setTextureMask(0, Shader::MaskFragment);
			pipeline.setColorFormat(window.getColorFormat());
			pipeline.setDepthFormat(window.getDepthFormat());
			pipeline.setMultisample(window.getMultisample());
			if(!pipeline.loadShaderGLSL(Shader::TypeVertex, "main.shader", "VERTEX_SHADER=1")) {
				TS_LOG(Error, "Application::create_resources(): can't load Vertex shader\n");
				return false;
			}
			if(!pipeline.loadShaderGLSL(Shader::TypeFragment, "main.shader", "FRAGMENT_SHADER=1")) {
				TS_LOG(Error, "Application::create_resources(): can't load Fragment shader\n");
				return false;
			}
			if(!pipeline.create()) {
				TS_LOG(Error, "Application::create_resources(): can't create Pipeline\n");
				return false;
			}
			
			// your resource creation code here
			
			return true;
		}
		
		// helper function to render the Window
		virtual bool render_window() {
			
			// update events
			Window::update();
			
			// update application
			if(!update()) return false;
			
			// render window
			if(!window.render()) return false;
			
			// compute target
			{
				Compute compute = device.createCompute();
				
				// dispatch application
				if(!dispatch(compute)) return false;
			}
			
			// flush surface texture
			device.flushTexture(surface);
			
			// window target
			target.begin();
			{
				Command command = device.createCommand(target);
				
				// render application
				if(!render(command)) {
					target.end();
					return false;
				}
			}
			target.end();
			
			// present window
			if(!window.present()) return false;
			
			// check errors
			if(!device.check()) return false;
			
			return true;
		}
		
		App app;				// Application interface
		Window window;			// Window interface
		Device device;			// Device interface
		Target target;			// Target interface
		
		Texture surface;		// Surface interface
		
		Kernel kernel;			// Kernel interface
		Pipeline pipeline;		// Pipeline interface
};

/*
 */
int32_t main(int32_t argc, char **argv) {
	
	// instantiate the application
	Application application(argc, argv);
	
	// create the application
	if(!application.create()) return 1;
	
	// run the application
	if(!application.run()) return 1;
	
	return 0;
}

/*
 */
#if _WIN32
	#include <system/TellusimWindows.h>
	TS_DECLARE_WINDOWS_HIGH_PERFORMANCE
#elif _ANDROID
	#include <system/TellusimAndroid.h>
	TS_DECLARE_ANDROID_NATIVE_ACTIVITY
#endif

File: tools/project/templates/python/app_core_compute.py
#!/usr/bin/env python3

# Tellusim Compute Application Template

import sys

from tellusimd import *

#
# main
#
def main(argv):
	
	app = None
	window = None
	device = None
	target = None
	
	surface = None
	
	kernel = None
	pipeline = None
	
	#
	# create function responsible for initialization
	#
	def create():
		
		nonlocal app
		
		# create app
		app = App(argv)
		if not app.create():
			Log.print(Log.Error, 'create(): can\'t create App\n')
			return False
		
		# create window
		if not create_window():
			return False
		
		# create resources
		if not create_resources():
			return False
		
		# additional initialization code can go here
		
		return True
	
	#
	# update function for logic updates
	#
	def update():
		
		# your update logic here
		
		return True
	
	#
	# dispatch function to handle the compute commands
	#
	def dispatch(compute):
		
		nonlocal surface
		
		# resize surface texture
		width = window.width
		height = window.height
		if not surface or surface.width != width or surface.height != height:
			if surface: device.releaseTexture(surface)
			surface = device.createTexture2D(FormatRGBAu8n, width, height, flags = Texture.FlagSurface)
			if not surface:
				Log.print(Log.Error, 'dispatch(): can\'t create Texture\n')
				return False
		
		# compute uniform parameters
		parameters = bytearray()
		parameters += Scalarf(Time.seconds())
		
		# dispatch kernel
		compute.setKernel(kernel)
		compute.setUniform(0, parameters)
		compute.setSurfaceTexture(0, surface)
		compute.dispatch(surface)
		compute.barrier(surface)
		
		# your dispatch code here
		
		return True
	
	#
	# render function to handle the drawing commands
	#
	def render(command):
		
		# draw fullscreen triangle
		command.setPipeline(pipeline)
		command.setTexture(0, surface)
		command.drawArrays(3)
		
		# your rendering code here
		
		return True
	
	#
	# helper function to create and initialize the Window
	#
	def create_window():
		
		nonlocal window, device, target
		
		# create Window
		window = Window(app.platform, app.device)
		if not window or not window.setSize(app.width, app.height) \
			or not window.create(window.platform_name + ' Python @NAME@') or not window.setHidden(False):
			Log.print(Log.Error, 'create_window(): can\'t create Window\n')
			return False
		
		# set Window callbacks
		def pressed_callback(key, code):
			if key == Window.KeyF4 and window.getKeyboardKey(Window.Key.Alt): window.stop()
			if key == Window.KeyEsc: window.stop()
		window.setKeyboardPressedCallback(pressed_callback)
		window.setCloseClickedCallback(lambda: window.stop())
		
		# create Device
		device = Device(window)
		if not device:
			Log.print(Log.Error, 'create_window(): can\'t create Device\n')
			return False
		
		# create Target
		target = device.createTarget(window)
		if not target:
			Log.print(Log.Error, 'create_window(): can\'t create Target\n')
			return False
		
		return True
	
	#
	# helper function to create and load resources
	#
	def create_resources():
		
		nonlocal kernel, pipeline
		
		# create kernel
		kernel = device.createKernel().setSurfaces(1).setUniforms(1)
		if not kernel.loadShaderGLSL("main.shader", "COMPUTE_SHADER=1"):
			Log.print(Log.Error, 'create_resources(): can\'t load Compute shader\n')
			return False
		if not kernel.create():
			Log.print(Log.Error, 'create_resources(): can\'t create Kernel\n')
			return False
		
		# create pipeline
		pipeline = device.createPipeline()
		pipeline.setTextureMask(0, Shader.MaskFragment)
		pipeline.setColorFormat(window.getColorFormat())
		pipeline.setDepthFormat(window.getDepthFormat())
		if not pipeline.loadShaderGLSL(Shader.TypeVertex, "main.shader", "VERTEX_SHADER=1"):
			Log.print(Log.Error, 'create_resources(): can\'t load Vertex shader\n')
			return False
		if not pipeline.loadShaderGLSL(Shader.TypeFragment, "main.shader", "FRAGMENT_SHADER=1"):
			Log.print(Log.Error, 'create_resources(): can\'t load Fragment shader\n')
			return False
		if not pipeline.create():
			Log.print(Log.Error, 'create_resources(): can\'t create Pipeline\n')
			return False
		
		# your resource creation code here
		
		return True
	
	#
	# helper function to render the Window
	#
	def render_window():
		
		# update events
		Window.update()
		
		# update application
		if not update(): return False
		
		# render window
		if not window.render(): return False
		
		# compute target
		if True:
			
			compute = device.createCompute()
			
			# dispatch application
			if not dispatch(compute): return False
			
			compute = None
		
		# flush surface texture
		device.flushTexture(surface)
		
		# window target
		if target.begin():
			
			command = device.createCommand(target)
			
			# render application
			if not render(command):
				target.end()
				return False
			
			command = None
			
			target.end()
		
		# present window
		if not window.present(): return False
		
		# check errors
		if not device.check(): return False
		
		return True
	
	# create the application
	if not create():
		return 1
	
	# run the application
	window.run(render_window)
	
	# finish context
	window.finish()
	
	return 0
	
#
# entry point
#
if __name__ == '__main__':
	try:
		exit(main(sys.argv))
	except Exception as error:
		print('\n' + str(error))
		exit(1)
